nrf_modem_lib: `_ON_INIT` and `_ON_SHUTDOWN` macros (#7283)
nrfconnect/sdk-nrf: samples/nrf9160/modem_callbacks/README.rst

.. _modem_callbacks_sample:

nRF9160: Modem callbacks
########################

.. contents::
   :local:
   :depth: 2

> The Modem callbacks sample demonstrates how to setup callbacks for Modem library initialization and shutdown calls, using the :ref:`nrf_modem_lib_readme` and how to setup a callback on changes to the modem function mode using the :ref:`lte_lc_readme` library.

Reviewer: ```suggestion
The Modem callbacks sample demonstrates how to setup callbacks for Modem library initialization and shutdown calls, using the :ref:`nrf_modem_lib_readme` and how to setup a callback on changes to the modem functional mode using the :ref:`lte_lc_readme` library.
```

The sample supports the following development kit:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf9160dk_nrf9160_ns

.. include:: /includes/spm.txt

Overview
********

The sample registers three callbacks at compile time: :c:func:`on_modem_init`, :c:func:`on_modem_shutdown` and :c:func:`on_cfun` using the :c:macro:`NRF_MODEM_LIB_ON_INIT`, :c:macro:`NRF_MODEM_LIB_ON_SHUTDOWN` and :c:macro:`LTE_LC_ON_CFUN` macros respectively.
Next, the sample initializes and shuts down the :ref:`nrfxlib:nrf_modem`, triggering :c:func:`on_modem_init` and :c:func:`on_modem_shutdown`.
Then, the sample initializes the modem library again, changes function mode using the :c:func:`lte_lc_func_mode_set` function in the :ref:`lte_lc_readme` library, and shuts down the modem library again.
This triggers the :c:func:`on_modem_init`, :c:func:`on_cfun` and :c:func:`on_modem_shutdown` callbacks.

Building and running
********************

.. |sample path| replace:: :file:`samples/nrf9160/modem_callbacks`

.. include:: /includes/build_and_run_nrf9160.txt

Testing
=======

|test_sample|

1. |connect_kit|
#. |connect_terminal|
#. Power on or reset your nRF9160 DK.

> #. Observe that the sample starts, creates and configures a PDP context, and then activates a PDN connection.

Reviewer: Is this PDP/PDN stuff leftovers from a different sample?

Sample output
=============

> The sample shows the following output, which may vary based on the network provider:

Reviewer: Probably the output shouldn't vary based on the network provider?

.. code-block:: console

  Modem callbacks sample started
  Initializing modem library
  > Initialized with value 0
  Shutting down modem library
  > Shutting down

>   Initializing modem library (again)

Reviewer 2: This (and the next lines) belongs to the [lib: lte_lc: add LTE_LC_ON_CFUN macro](https://github.com/nrfconnect/sdk-nrf/pull/7283/commits/27b48f723b65d7f7067c0bcb3ac672348f5d05e8) commit.

  > Initialized with value 0

>   Changing function mode
>   > Function mode has changed to 1

Reviewer: ```suggestion
  Changing functional mode
  > Functional mode has changed to 1
```

  Shutting down modem library (again)

>   > Function mode has changed to 0

Reviewer: ```suggestion
  > Functional mode has changed to 0
```

.. _modem_callbacks_sample:

nRF9160: Modem callbacks
########################

.. contents::
   :local:
   :depth: 2

The Modem callbacks sample demonstrates how to setup callbacks for Modem library initialization and shutdown calls, using the :ref:`nrf_modem_lib_readme` and how to setup a callback for changes to the modem functional mode using the :ref:`lte_lc_readme` library.

Requirements
************

The sample supports the following development kit:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf9160dk_nrf9160_ns

.. include:: /includes/spm.txt

Overview
********

> The sample registers three callbacks at compile time: :c:func:`on_modem_init`, :c:func:`on_modem_shutdown` and :c:func:`on_cfun` using the :c:macro:`NRF_MODEM_LIB_ON_INIT`, :c:macro:`NRF_MODEM_LIB_ON_SHUTDOWN` and :c:macro:`LTE_LC_ON_CFUN` macros respectively.
> Next, the sample initializes and shuts down the :ref:`nrfxlib:nrf_modem`, triggering :c:func:`on_modem_init` and :c:func:`on_modem_shutdown`.
> Then, the sample initializes the modem library again, changes functional mode using the :c:func:`lte_lc_func_mode_set` function in the :ref:`lte_lc_readme` library, and shuts down the modem library again.
> This triggers the :c:func:`on_modem_init`, :c:func:`on_cfun` and :c:func:`on_modem_shutdown` callbacks.

Reviewer 3: ```suggestion
The sample performs the following operations:

1. Registers callbacks during compile time for :c:func:`on_modem_init`, :c:func:`on_modem_shutdown` and :c:func:`on_cfun` functions using the :c:macro:`NRF_MODEM_LIB_ON_INIT`, :c:macro:`NRF_MODEM_LIB_ON_SHUTDOWN` and :c:macro:`LTE_LC_ON_CFUN` macros respectively.
#. Initializes and shuts down the :ref:`nrfxlib:nrf_modem` by triggering :c:func:`on_modem_init` and :c:func:`on_modem_shutdown` functions.
#. Initializes the modem library.
#. Changes functional mode using the :c:func:`lte_lc_func_mode_set` function in the :ref:`lte_lc_readme` library
#. Shuts down the modem library.

This triggers the callbacks for :c:func:`on_modem_init`, :c:func:`on_cfun` and :c:func:`on_modem_shutdown` functions.
```

Building and running
********************

.. |sample path| replace:: :file:`samples/nrf9160/modem_callbacks`

.. include:: /includes/build_and_run_nrf9160.txt

Testing
=======

|test_sample|

1. |connect_kit|
#. |connect_terminal|
#. Power on or reset your nRF9160 DK.
#. Observe that the sample starts, initializes and shuts down the modem.

> #. Observe that the sample initializes the modem, changes functional modem and shuts down the modem.

Reviewer: ```suggestion
#. Observe that the sample initializes the modem, changes functional mode and shuts down the modem.
```
